From 5fccd58b35f4a4536d6aacc055bb4246f322f880 Mon Sep 17 00:00:00 2001
From: Martin Valgur <martin.valgur@gmail.com>
Date: Sun, 8 Sep 2024 22:23:16 +0300
Subject: [PATCH] Use OBJECT internal libs instead of fragile merge_libraries()

merge_libraries() also merges all external static Conan dependencies into the library, which is unwanted in most cases.
This causes AppleClang builds to fail due to duplicate symbols in the final library, for instance.

---
 CMakeLists.txt                | 4 ++--
 cdk/core/CMakeLists.txt       | 2 +-
 cdk/foundation/CMakeLists.txt | 2 +-
 cdk/mysqlx/CMakeLists.txt          | 2 +-
 cdk/parser/CMakeLists.txt     | 2 +-
 cdk/protocol/mysqlx/CMakeLists.txt | 2 +-
 common/CMakeLists.txt         | 2 +-
 devapi/CMakeLists.txt         | 2 +-
 xapi/CMakeLists.txt           | 2 +-
 9 files changed, 10 insertions(+), 10 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 341ed2de..5a55caff 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -409,8 +409,8 @@ add_subdirectory(devapi)
 #
 
 # Generate the main connector library.
-
-merge_libraries(connector xapi devapi)
+add_library(connector cmake/libutils/empty.cc)
+target_link_libraries(connector PUBLIC cdk cdk_foundation cdk_mysqlx cdk_parser cdk_proto_mysqlx common devapi xapi)
 add_version_info(connector
   "MySQL Connector/C++ XDevAPI library."
   "Implements MySQL Connector/C++ XDevAPI."
diff --git a/cdk/core/CMakeLists.txt b/cdk/core/CMakeLists.txt
index ba7ee3e0..25a48433 100644
--- a/cdk/core/CMakeLists.txt
+++ b/cdk/core/CMakeLists.txt
@@ -41,4 +41,4 @@ SET(cdk_sources
 
 file(GLOB HEADERS *.h)
 
-add_library(cdk STATIC ${cdk_sources} ${HEADERS})
+add_library(cdk OBJECT ${cdk_sources} ${HEADERS})
diff --git a/cdk/foundation/CMakeLists.txt b/cdk/foundation/CMakeLists.txt
index 3cd3bfc4..778516f4 100644
--- a/cdk/foundation/CMakeLists.txt
+++ b/cdk/foundation/CMakeLists.txt
@@ -52,7 +52,7 @@ file(GLOB HEADERS *.h)
 
 #message("cdk foundation sources: ${sources}")
 
-add_library(cdk_foundation STATIC ${sources} ${HEADERS})
+add_library(cdk_foundation OBJECT ${sources} ${HEADERS})
 add_coverage(cdk_foundation)
 
 target_include_directories(cdk_foundation PUBLIC
diff --git a/cdk/mysqlx/CMakeLists.txt b/cdk/mysqlx/CMakeLists.txt
index aa358f03..af48bb11 100644
--- a/cdk/mysqlx/CMakeLists.txt
+++ b/cdk/mysqlx/CMakeLists.txt
@@ -41,7 +41,7 @@ endif()
 
 file(GLOB HEADERS *.h)
 
-ADD_LIBRARY(cdk_mysqlx STATIC
+ADD_LIBRARY(cdk_mysqlx OBJECT
   session.cc
   result.cc
   auth_hash.cc
diff --git a/cdk/parser/CMakeLists.txt b/cdk/parser/CMakeLists.txt
index 5e1a4bd8..4e5e1b49 100644
--- a/cdk/parser/CMakeLists.txt
+++ b/cdk/parser/CMakeLists.txt
@@ -31,7 +31,7 @@ ADD_SUBDIRECTORY(tests)
 
 file(GLOB HEADERS *.h)
 
-add_library(cdk_parser STATIC
+add_library(cdk_parser OBJECT
   tokenizer.cc
   json_parser.cc
   expr_parser.cc
diff --git a/cdk/protocol/mysqlx/CMakeLists.txt b/cdk/protocol/mysqlx/CMakeLists.txt
index a9873127..21f1a8a2 100644
--- a/cdk/protocol/mysqlx/CMakeLists.txt
+++ b/cdk/protocol/mysqlx/CMakeLists.txt
@@ -116,7 +116,7 @@ mysqlx_protobuf_generate_cpp(PB_SRCS PB_HDRS ${proto_mysqlx_defs})
 
 file(GLOB HEADERS *.h)
 
-ADD_LIBRARY(cdk_proto_mysqlx STATIC
+ADD_LIBRARY(cdk_proto_mysqlx OBJECT
             protocol.cc protocol_compression.cc session.cc rset.cc
             stmt.cc crud.cc
             ${PB_SRCS}
diff --git a/common/CMakeLists.txt b/common/CMakeLists.txt
index 587340e3..fec8ae01 100644
--- a/common/CMakeLists.txt
+++ b/common/CMakeLists.txt
@@ -29,7 +29,7 @@
 
 file(GLOB HEADERS *.h)
 
-add_library(common STATIC
+add_library(common OBJECT
   session.cc result.cc collection.cc value.cc
   ${HEADERS}
 )
diff --git a/devapi/CMakeLists.txt b/devapi/CMakeLists.txt
index e6982d2e..32f60a7c 100644
--- a/devapi/CMakeLists.txt
+++ b/devapi/CMakeLists.txt
@@ -32,7 +32,7 @@
 
 file(GLOB HEADERS *.h)
 
-add_library(devapi STATIC
+add_library(devapi OBJECT
   session.cc
   result.cc
   document.cc
diff --git a/xapi/CMakeLists.txt b/xapi/CMakeLists.txt
index 25f3e579..e73bb334 100644
--- a/xapi/CMakeLists.txt
+++ b/xapi/CMakeLists.txt
@@ -53,7 +53,7 @@ set(_libmysqlx_cc_src
 
 file(GLOB HEADERS *.h)
 
-add_library(xapi STATIC ${_libmysqlx_cc_src} ${HEADERS})
+add_library(xapi OBJECT ${_libmysqlx_cc_src} ${HEADERS})
 target_link_libraries(xapi PUBLIC common)
 add_coverage(xapi)
 
